XML support

RikkiGibson opened the pull request:
- Adds xml2js
- Adds metadata to mappers for XML serialization (name, elementName, isWrapped, etc)
- Adds extra parsing behaviors for booleans and numbers. It's ambiguous to xml2js what kind of value conversion should take place, so we have to do it.
  - xml2js has "attribute value converters" for numbers and booleans that can be applied out of the box, but I would prefer to only do it if the Swagger says it should happen.

To do:
- [ ] ~~Handle model flattening~~ (seems unreasonably difficult because the AutoRest SerializedName doesn't take the XmlName into account)
  - PUT storage account with attributes, child elements, wrapped lists, embedded lists
- [x] Handle all review feedback asks
- [x] xml2js attrkey
- [x] port latest from azure sdk for node #9

RikkiGibson commented on lib/serializer.ts:
```diff
@@ -310,38 +311,58 @@ export class Serializer {
-        if (modelProps.hasOwnProperty(key)) {
-          const paths = this.splitSerializeName(modelProps[key].serializedName);
-          const propName = paths.pop();
+      for (const key of Object.keys(modelProps)) {
```
```js
for (const key of Object.keys(modelProps)) { ... }
```
is equivalent to
```js
for (const key in modelProps) {
  if (modelProps.hasOwnProperty(key) {
    ...
  }
}
```

daschult replied:
Oh, weird. Okay.

RikkiGibson commented on lib/serializer.ts:
```diff
@@ -310,38 +311,58 @@ export class Serializer {
             }
             parentObject = parentObject[pathName];
           }
+        }
```
Mainly I struggled to decide what parameter flattening means in XML. Maybe we can reason that out, or decide whether we're even supporting the feature or not.

RikkiGibson commented on lib/serializer.ts:
```diff
@@ -512,13 +558,46 @@ export class Serializer {
+        // so let's do the more user-friendly thing and return an empty list.
+        responseBody = [];
+      } else {
+        return responseBody;
```
So, this is a reasonable code structure question. This used to be a central "if the thing ain't there, don't try to map it" point. But there's now an edge case just above it that makes it a little less clear. I first tried to work on this by making every deserialization "helper" be resilient to undefined/null values (currently some are and others are not), but it was so much more copy-pasting of this kind of check that I eventually just went with this.

RikkiGibson commented on lib/serializer.ts:
```diff
@@ -310,38 +311,58 @@ export class Serializer {
+            propName = propertyMapper.xmlElementName || propertyMapper.xmlName;
+          }
+        } else {
+          const paths = this.splitSerializeName(propertyMapper.serializedName);
```
Good question--looks like Dan wondered the same thing. I think I just glossed over it because the way the property name is obtained for XML is different than in JSON. We could maybe just take the propName extracted on line 323, and pass it to splitSerializeName, and carry on from there.

Note that the C# runtime does not handle parameter flattening for XML and we aren't aware of any services that would need it for XML.

amarzavery commented on lib/serializer.ts:
```diff
@@ -512,13 +558,46 @@ export class Serializer {
    */
   deserialize(mapper: Mapper, responseBody: any, objectName: string): any {
-    if (responseBody === null || responseBody === undefined) return responseBody;
+    if (responseBody == undefined) {
```
=== is different than ==

amarzavery replied:
and why remove the responseBody === null check?

RikkiGibson replied:
This is an intentional style change based on `foo === undefined || foo === null` being equivalent to `foo == undefined`.

Your input on which style to go with is welcomed--we can configure tslint to only allow `==` in the case of `expr == undefined` comparison.

amarzavery replied:
By not having the null check this means that we want to process the responseBody when it is null. Is there are reason to do so?

amarzavery replied:
ok learnt something new today. Played in the node REPL.
```
> null == undefined
true
> null === undefined
false
>
```

amarzavery commented on lib/serializer.ts:
```diff
@@ -512,13 +558,46 @@ export class Serializer {
+        responseBody = [];
+      } else {
+        return responseBody;
+      }
```
Isn't the else part redundant ? The outer return responseBody will eventually happen any ways.

RikkiGibson replied:
I think Dan brought this up, see https://github.com/Azure/ms-rest-js/pull/19#discussion_r178192141

amarzavery commented on lib/serializer.ts:
```diff
@@ -435,8 +457,30 @@ export class Serializer {
+        const propertyMapper = modelProps[key];
+        let propertyObjectName = objectName;
+        if (propertyMapper.serializedName !== "") {
+          propertyObjectName = objectName + "." + modelProps[key].serializedName;
```
modelProps[key] -> propertyMapper (just for consistency)

RikkiGibson replied:
Done

amarzavery commented on lib/serializer.ts:
```diff
@@ -5,11 +5,12 @@ import * as utils from "./util/utils";
 import { duration, isDuration } from "moment";
 const isBuffer = require("is-buffer");
 import * as isStream from "is-stream";
+import { isArray } from "util";
```
We don't need this. Be vary of importing the util library from node.js This has a direct impact on the bundle size in the browser. There is a function on the Array type `Array.isArray()`, use that instead.

RikkiGibson replied:
Looking into removing all the "util" usages--would really like a promisify alternative though.

RikkiGibson replied:
Looks like it's simply `new Promise(function(resolve, reject) { ...})` :)

amarzavery replied:
We have never used node.js's util library at all. 
Note: We have a util folder in the runtime and importing stuff from there is fine. Just making sure there is no confusion.
And why would we ever need a Promisified version of `Array.isArray()` ?

amarzavery replied:
`Array.isArray()` is a synchronous call.

RikkiGibson replied:
I was making use of `promisify` from within the util module in another file :)

amarzavery replied:
Ahh. I see. Yup that is super simple.
```
return new Promise((resolve, reject) => {
  if (err) reject(err);

  resolve(result);
});
```

That is what we are essentially doing in the generated code when we say we support callback and promises both [for example](https://github.com/Azure/azure-sdk-for-node/blob/master/lib/services/storageManagement2/lib/operations/storageAccounts.js#L2994-L3000).

amarzavery replied:
It is better to use arrow functions `() => {}` over `function () {}` since the `this` is preserved inside the arrow function

```javascript
function foo() {
  this.name = "foo";
   const bar = () => {
      console.log(this.name); // foo
   };
}
```

amarzavery replied:
v/s 
```javascript
function foo() {
  this.name = "foo";
  let self = this;
   function bar() => {
      console.log(this.name); // undefined
      console.log(self.name); //"foo"
   };
}
```

RikkiGibson commented on lib/serializer.ts:
```diff
@@ -310,38 +311,58 @@ export class Serializer {
-            continue;
+        // make sure required properties of the CompositeType are present
+        if (modelProps[key].required && !modelProps[key].isConstant) {
+          if (object[key] === null || object[key] === undefined) {
```
Looking at it a bit more, I'm less enthused about the code churn, but could consider either changing all usages of `foo === null || foo === undefined` to `foo == undefined` in a different PR. Just don't want to create noise with that.

RikkiGibson commented on lib/serializer.ts:
```diff
@@ -416,7 +437,8 @@ export class Serializer {
     };
-    if (responseBody !== null && responseBody !== undefined) {
+    responseBody = responseBody || {};
+    if (responseBody) {
```
Done

RikkiGibson commented on lib/httpOperationResponse.ts:
```diff
@@ -22,11 +22,12 @@ export class HttpOperationResponse {
+   * The response body as parsed JSON or XML
    */
-  bodyAsJson: { [key: string]: any } | Array<any> | string | number | boolean | null | void;
+  bodyAsJson?: { [key: string]: any } | Array<any> | string | number | boolean | null | void;
```
This also requires a change in ms-rest-azure-js, so I'll open a PR there pretty soon.

RikkiGibson commented on package.json:
```diff
@@ -30,29 +30,31 @@
   "license": "MIT",
```
Bumped to 0.3.0

RikkiGibson commented:
I think I'm missing a bit of test coverage: an element with attributes and a simple text body like the following:

```xml
<pear color="green">Bartlett</pear>
```
I actually don't even know how to represent this data in a Swagger! An element with an attribute typically corresponds to an object type, where the property "color" would have "xmlIsAttribute = true", but there is no property name you can keep the text content under.

@lmazuel @olydis any idea how this would be represented?

olydis commented:
@RikkiGibson that's not representable for the very reason you state AFAIK - that was also one of the things we stumbled across for storage, I guess without an extension (I don't know, maybe to mark a certain property as the "body"? But that implies that there can be no other non-attribute properties... ughhhhh) that wouldn't work. But nothing in that direction happened from our side, do you see a need for that?

RikkiGibson commented:
Seems like this is a limitation of [OpenAPI 2.0 itself](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#xmlObject) hitting us and we should just let it lie until an important enough service starts asking for it.

RikkiGibson commented:
Addressed review feedback-- I'm shelving XML flattening for now because the serializedName doesn't give us the XmlNames that we need to traverse the parsed document.
